src/runtime/hexagon/hexagon_vtcm_pool.cc

```diff
@@ -0,0 +1,166 @@
+  // hanging, both in the simulator and on hardware.
+  context_id_ = HAP_compute_res_acquire(&res_info, /*timeout = */ 0);
+  CHECK(context_id_) << "HAP_compute_res_acquire failed to acquire requested VTCM resource.";
+  HEXAGON_SAFE_CALL(HAP_compute_res_attr_get_vtcm_ptr_v2(&res_info, &vtcm_data_, &vtcm_size_));
+  CHECK(vtcm_data_ != nullptr) << "HAP_compute_res_acquire returned nullptr when allocating VTCM.";
+  CHECK(vtcm_size_ >= min_size)
+      << "HAP_compute_res_acquire failed to allocate minimum amount of VTCM";
+  free_.emplace_back(std::pair<char*, size_t>(static_cast<char*>(vtcm_data_), vtcm_size_));
+  // DebugDump();
+}
+
+HexagonVtcmPool::~HexagonVtcmPool() { HEXAGON_SAFE_CALL(HAP_compute_res_release(context_id_)); }
+
+void* HexagonVtcmPool::Allocate(size_t nbytes) {
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  CHECK(!free_.empty()) << "No free VTCM";
+
+  // If this is not aligned on a 2k block, allocate from the end to avoid fragmentation
+  if (nbytes & size_t(0x7FF)) {
+    DLOG(INFO) << "VTCM nbytes requested: " << nbytes << " allocate from the end";
+    auto last_free_entry = free_.rbegin();
+    CHECK(last_free_entry->second >= nbytes)
+        << "Not enough contiguous VTCM space at the end to allocate";
+    char* ptr = last_free_entry->first + (last_free_entry->second - nbytes);
+    allocations_.emplace_back(std::pair<char*, size_t>(ptr, nbytes));
+    last_free_entry->second -= nbytes;
+    // DebugDump();
+    return ptr;
+  }
+
+  std::pair<char*, size_t>& entry_to_allocate = free_.front();
+  for (auto entry : free_) {
+    if ((entry.second < entry_to_allocate.second) && (entry.second >= nbytes)) {
+      entry_to_allocate = entry;
+      if (entry_to_allocate.second == nbytes) {
+        break;
+      }
+    }
+  }
+  CHECK(entry_to_allocate.second >= nbytes) << "Not enough contiguous VTCM space to allocate";
+  char* ptr = entry_to_allocate.first;
+  allocations_.emplace(allocations_.end(), std::pair<char*, size_t>(ptr, nbytes));
+
+  for (auto it = free_.begin(); it != free_.end(); it++) {
+    if (ptr == it->first) {
+      if (it->second == nbytes) {
+        free_.erase(it);
+      } else {
+        it->first = it->first + nbytes;
+        it->second = it->second - nbytes;
+      }
+      break;
+    }
+  }
+  // DebugDump();
+  return ptr;
+}
+
+void HexagonVtcmPool::Free(void* ptr, size_t nbytes) {
+  char* ptr_to_free = static_cast<char*>(ptr);
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  bool found_allocation_entry = false;
+  for (auto it = allocations_.begin(); it != allocations_.end(); it++) {
+    if (ptr_to_free == it->first) {
+      CHECK(it->second == nbytes) << "Attempted to free a different size than was allocated";
+      allocations_.erase(it);
+      found_allocation_entry = true;
+      break;
+    }
+  }
+  CHECK(found_allocation_entry) << "Attempted to free a pointer that had not been allocated";
+
+  auto it = free_.begin();
+  for (; it != free_.end(); it++) {
```

Comment:
You could use `std::lower_bound` (and move back by one) instead of manual search.

Reply:
This is iterating over a different list now.  (First I remove the entry from allocations_, then adjust free_ to incorporate the freed space.)

Reply:
I'm talking about something like this
```
std::lower_bound(free_.begin(), free_.end(), std::pair<char*, size_t>(ptr_to_free, nbytes),
                 [](auto p, auto q) { return p.first < q.first; });
```

Reply:
Thanks, done